userspace/engine/rule_loader.h

```diff
@@ -33,110 +33,207 @@ class falco_engine;
-		extra string is appended at the end of the rule output. If a rule
-		output does not contain "%container.info", then this flag has no effect
-		and the extra string is appended at the end of the rule output anyways.
+		\brief Represents infos about a macro 
 	*/
-	virtual void configure(
-		falco_common::priority_type min_priority,
-		bool replace_container_info,
-		const std::string& extra);
+	struct macro_info
+	{
+		mark context;
+		bool used;
+		size_t index;
+		size_t visibility;
+		std::string name;
+		std::string cond;
+		std::string source;
+		std::shared_ptr<libsinsp::filter::ast::expr> cond_ast;
+	};
+
+	/*!
+		\brief Represents infos about a single rule exception
+	*/
+	struct rule_exception_info
+	{
+		/*!
+			\brief This is necessary due to the dynamic-typed nature of
+			exceptions. Each of fields, comps, and values, can either be a
+			single value or a list of values. This is a simple hack to make
+			this easier to implement in C++, that is not non-dynamic-typed.
+		*/
+		struct entry {
+			bool is_list;
+			std::string item;
+			std::vector<entry> items;
+
+			inline bool is_valid() const
+			{
+				return (is_list && !items.empty())
+					|| (!is_list && !item.empty());
+			}
+		};
+
+		std::string name;
+		entry fields;
+		entry comps;
+		std::vector<entry> values;
+	};
+
+	/*!
+		\brief Represents infos about a rule 
+	*/
+	struct rule_info
+	{
+		mark context;
+		size_t index;
+		size_t visibility;
+		std::string name;
+		std::string cond;
+		std::string source;
+		std::string desc;
+		std::string output;
+		std::set<std::string> tags;
+		std::vector<rule_exception_info> exceptions;
+		falco_common::priority_type priority;
+		bool enabled;
+		bool warn_evttypes;
+		bool skip_if_unknown_filter;
+	};
+
+	/*!
+		\brief Erases all the internal state and definitions
+	*/
+	virtual void clear();
 
 	/*!
 		\brief Returns true if the given plugin name and version are compatible
-		with the loaded rulesets. If false is returned, required_version is
+		with the internal definitions. If false is returned, required_version is
 		filled with the required plugin version that didn't match.
 	*/
 	virtual bool is_plugin_compatible(
 		const std::string& name,
 		const std::string& version,
 		std::string& required_version);
+	
+	/*!
+		\brief Uses the internal state to compile a list of falco_rules
+	*/
+	bool compile(context& ctx, indexed_vector<falco_rule>& out);
+
+	/*!
+		\brief Defines an info block. If a similar info block is found
+		in the internal state (e.g. another rule with same name), then
+		the previous definition gets overwritten
+	*/
+	virtual void define(context& ctx, engine_version_info& info);
+	virtual void define(context& ctx, plugin_version_info& info);
+	virtual void define(context& ctx, list_info& info);
+	virtual void define(context& ctx, macro_info& info);
+	virtual void define(context& ctx, rule_info& info);
+
+	/*!
+		\brief Appends an info block to an existing one. An exception
```

Comment:
We are definitely not consistent about error reporting, but I'd like to make error reporting more consistent, and get away from using exceptions, if possible. I hope that follow-on PRs can provide a richer "error" struct that points to the rule/macro/list that had the error + an error string + a filename etc. I think it's hard to put that kind of info in an exception what().

I know you already used exceptions a lot in the previous iteration of this, and exceptions are still used a lot in rule_reader, so I'm happy to do a separate pass if we can get consensus about how to return these kinds of errors.

Reply:
I think this is a broader topic and we should go for it in a second pass, but I agree that having a well-defined error reporting mechanism would be super helpful. First ideas on top of my mind:
- We can create an "error manager" class to which errors and warnings are reported (instead of having the two vectors) and which decides when an error is blocking/fatal by also gathering as much metadata as possible
- We can create a "falco rule loader exception" that extends `falco_exception` but contains all the info you mentioned

I would personally lean towards option 1 because it allows adding more logic in the error reporting tasks. Plus, exceptions are always blocking in the rule loading logic, and are not suitable for warnings. If we had a class for this, we can potentially make it the single source of truth for loading metadata too, so that we don't need to store all the 'context' information in each rule_info/macro_info/list_info.

Reply:
Sure, we can work on a design and then fix it later.